src/typed-ast/expr.h

```diff
@@ -0,0 +1,637 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "typed-ast/node.h"
+#include "typed-ast/print.h"
+
+#include <memory>
+#include <optional>
+#include <vector>
+
+namespace verona::ast
+{
+  class MemberDef;
+
+  /// This is the base class for expressions, from which all concrete nodes
+  /// derive.
+  class Expr : public Node
+  {
+  public:
+    enum class Kind
+    {
+      LocalDecl,
+      LocalRef,
+      MemberRef,
+      Assignment,
+      Sequence,
+      If,
+      While,
+      Continue,
+      Return,
+      Yield,
+      Break,
+      Invoke,
+      MethodCall,
+      StaticCall,
+      IntegerLiteral,
+      FloatLiteral,
+      BooleanLiteral,
+      StringLiteral,
+      Interpolate,
+      Tuple,
+      When,
+      New,
+      Lambda,
+    };
+
+    Kind getKind() const
+    {
+      return kind;
+    }
+
+  protected:
+    explicit Expr(Kind kind, SourceLocation location)
+    : Node(location), kind(kind)
+    {}
+    Kind kind;
+  };
+
+  using ExprPtr = std::unique_ptr<Expr>;
+
+  /// This represents a local variable declaration, e.g. `let x`.
+  /// TODO: unclear what this "evaluates" to.
```

Comment:
In the generator, this is only used to fetch the name for the symbol table and the type for overriding in an assignment.

ex:
```
let a : U32 = 10;
let b : U16 = a;
```

When assigning `b` we add a `verona.cast -> U16` and set that as the new symbol's SSA value.

Reply:
Well the question is, if `let x` is as good an expression as anything else, does `(let x).foo()` mean anything? What about `(let x = E).foo()`?

We probably want to introduce a verification pass that checks that `let x` does not appear in an rvalue context. Therefore the question "what does it evaluate to", in an rvalue context, is meaningless.